Source/HelixToolkit.Wpf.Shared/Helpers/Viewport3DHelper.cs

```diff
@@ -223,49 +223,143 @@ public static IEnumerable<RectangleHitResult> FindHits(this Viewport3D viewport,
+
+                            Point3D[] point3dCorners = new Point3D[8] { p0, p1, p2, p3, p4, p5, p6, p7 };
+                            meshPoints= TransformPoints3dTo2d(viewport, transform, point3dCorners);
                         }
-
-                        if (mode == SelectionHitMode.Touch && status)
+                        else //MeshGeometry3D maybe (high probability) is changed from origin model3D (like is cutted...)
                         {
-                            break;
+                            //Get all point in actual mesh position 
+                            int length = distinctTriangleIndices.Length;
+                            meshPoints = new Point[length];
+                            for (var i = 0; i < length; i++)
+                            {
+                                int triangleIndice = distinctTriangleIndices[i];
+                                meshPoints[i] = point2Ds[triangleIndice];
+                            }
                         }
-                    }
+                        Rect geometryBound = GetBound2d(meshPoints);
 
-                    if (status)
-                    {
-                        results.Add(new RectangleHitResult(model, visual));
+                        if (rectangle.Contains(geometryBound))// object is both inside and touch
+                        {
+                            results.Add(new RectangleHitResult(model, visual));
+                        }
+                        else if (mode == SelectionHitMode.Touch && rectangle.IntersectsWith(geometryBound))
+                        {
+                            // evaluate each triangle
+                            int triangleIndiceCount = geometry.TriangleIndices.Count;
+                            for (var i = 0; i < triangleIndiceCount; i += 3)
+                            {
+                                var p1 = point2Ds[geometry.TriangleIndices[i]];
+                                var p2 = point2Ds[geometry.TriangleIndices[i + 1]];
+                                var p3 = point2Ds[geometry.TriangleIndices[i + 2]];
+                                var triangle = new Triangle(p1, p2, p3);
+                                // check if triangle touched/inside with rectangle
+                                if (triangle.IsCompletelyInside(rectangle)
+                                         || triangle.IntersectsWith(rectangle)
+                                         || triangle.IsRectCompletelyInside(rectangle))
+                                {
+                                    results.Add(new RectangleHitResult(model, visual));
+                                    break;
+                                }
+                            }
+                        }
                     }
                 });
-
             return results;
+            Point[] TransformPoints3dTo2d(Viewport3D viewport3d, Transform3D transform3d, Point3D[] point3ds)
```

Comment:
It's better to use `IList<Point3D>` as parameter instead of `Point3D[]` array, otherwise you have to create an array just for passing in the position list.

Reply:
Do you have any performance benchmark for comparison? Also, creating a temporary array and copy everything increases memory usage on large models, I don't think the cost can be justified. What happens if you have a mesh contains 1 million vertices?